### PR TITLE
feat: Add Async Sentry Transport

### PR DESCRIPTION
**Pull request by AbhiPrasad**
Dependent on #5 

#### Note: should only be merged after `sentry-go` changes

<details>
<summary> This PR has been reworked. For reference I have included the original PR description below</summary>

## Motivation

The first implementation of the transport in #5 was done just to get it working. Now that we determined it works, this PR aims to improve the transport by leveraging goroutines and a buffer to send requests.

## Implementation

Much of the implementation is based on how `sentry-go` does transport, but I use just a wait group because we only will flush on exporter shutdown.

In addition, I instrument the errors in the application by initializing Sentry with the given DSN.

## Testing

I wrote tests based on the functions added. I am not quite sure how to approach testing the `SendTransaction` function, so I would appreciate some help with that.

</details>

## Motivation

The current implementation of the collector relies quite a bit on replicating a bunch of features from the SDK. This PR aims to clean that up.

### CODE REVIEW

File: exporter/sentryexporter/transport.go [hunk 1]
```diff
@@ -110,25 +130,102 @@ func (t *SentryTransport) SendTransaction(transaction *SentryTransaction) error
...
+		return time.Second * time.Duration(seconds)
+	}
+
+	return defaultRetryAfter
+}
+
+func envelopeAPIURL(DSN *sentry.Dsn) string {
+	url := DSN.StoreAPIURL()
+	r := regexp.MustCompile("(/store/)")
+	url.Path = r.ReplaceAllString(url.Path, "/envelope/")
```
Comment (rhcarvalho) [inline exporter/sentryexporter/transport.go:229]: _Pro-tip not limited to this PR_

Regular expressions are cool; regular expressions are powerful. But they are not the solution to everything. And they are certainly not the best solution in many cases. They are also often easy to mess up.

In a case like this here, a simple string replacement would do:

```go
strings.Replace("https://domain/store/", "/store/", "/envelope/", -1)
```

Note that both the code above and the regexp solution are brittle and may produce the wrong output for valid inputs: https://play.golang.org/p/BQcA259wZ6B

While the regexp can be "fixed" / "improved" the point here is that it is the wrong tool for the job. It consumes a lot a extra resources unnecessarily, both from the computer and from humans.

---

And, of course, we don't expect this hack to produce the API endpoint in the final version of this :)
Wanted to take the opportunity to illustrate when NOT to use regexps.

**Comment by rhcarvalho** [comment]
> The first implementation of the transport in #5 was done just to get it working.

I think replicating a "transport" should be out of scope for the Exporter.
It should simply use the SDK.

Thoughts?

**Comment by AbhiPrasad** [comment]
Yes I completely agree the exporter should just use the SDK. 

Not sure what the timeline is for transaction support in `sentry-go`, so figured just porting the code for now is good enough. I can't just use `sentry.HTTPTransport` as the fields on that struct are private, so there is no way for me to interact with the buffer to add new requests.

Do you think we should add the needed structs and methods to `sentry-go` first, then come back to the collector?

File: exporter/sentryexporter/transport.go (hunk 1 above)
Comment (AbhiPrasad) [reply exporter/sentryexporter/transport.go]: Good point with the regex there, fixed.

**Comment by rhcarvalho** [comment]
> Do you think we should add the needed structs and methods to `sentry-go` first, then come back to the collector?

My original thought was to get it working first, and once we're happy we move it to `sentry-go`. Thus no need to reimplement queueing behavior etc.

If you feel you know what needs to change in the SDK, then let's start moving things there.

If you want to experiment with changes quick outside of the upstream SDK, you can use `go mod vendor` and patch the vendored SDK, or use a `replace` directive in `go.mod` to point the SDK to a local checkout of the SDK where you work your changes.

**Comment by AbhiPrasad** [comment]
Closing this PR for now while I make the sentry-go changes

**Pull request closed without merging**